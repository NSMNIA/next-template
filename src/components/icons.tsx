import type { Icon } from "@/types"
import {
  ArrowLeft,
  ArrowRight,
  Check,
  ChevronDown,
  ChevronLeft,
  ChevronRight,
  ChevronUp,
  Minus,
  Plus,
  X,
} from "lucide-react"

export const Icons = {
  logo: (props: Icon) => (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      fill="currentColor"
      viewBox="0 0 28 37"
      {...props}
    >
      <path d="M17.13 8.22243h-1.5988V23.7533H17.13c1.0003 0 1.5987-.5581 1.5987-2.1112V10.3337c0-1.55311-.5984-2.11127-1.5987-2.11127Z" />
      <path d="M8.67851 10.5289v1.7223c0 1.7922.75455 2.3663 1.99019 2.3663h1.2073V8.22243h-1.5091c-1.17434 0-1.68839.70393-1.68839 2.30647Z" />
      <path d="M13.7037 0 0 2.74486V28.8864L13.7037 37l13.7037-8.1136V2.74486L13.7037 0Zm8.5721 21.6297c0 3.4662-1.6814 5.3542-5.0976 5.3542h-5.2191v-9.2851h-1.1808c-1.55504 0-2.14304.6487-2.14304 2.6626v6.6225H5.23367v-6.4386c0-2.4205.55881-4.0519 2.26937-4.5796v-.0629c-1.52104-.6197-2.23537-2.0429-2.23537-4.2698v-1.3313c0-3.34514 1.521-4.98624 5.06353-4.98624h6.847c3.4017 0 5.0976 1.89284 5.0976 5.35904v10.9552Z" />
    </svg>
  ),
  close: X,
  chevronLeft: ChevronLeft,
  chevronRight: ChevronRight,
  chevronDown: ChevronDown,
  chevronUp: ChevronUp,
  arrowRight: ArrowRight,
  arrowLeft: ArrowLeft,
  check: Check,
  minus: Minus,
  plus: Plus,
}
