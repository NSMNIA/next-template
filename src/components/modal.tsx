"use client"

import * as React from "react"

import { useMediaQuery } from "@/hooks/use-media-query"
import {
  Dialog,
  DialogContent,
  DialogDescription,
  DialogHeader,
  DialogTitle,
  DialogTrigger,
} from "@/components/ui/dialog"
import {
  Drawer,
  DrawerContent,
  DrawerDescription,
  DrawerFooter,
  DrawerHeader,
  DrawerTitle,
  DrawerTrigger,
} from "@/components/ui/drawer"

interface ModalProps {
  trigger: React.ReactNode
  title: string | React.ReactNode
  description: string | React.ReactNode
}

export function Modal({
  trigger,
  title,
  description,
  children,
}: Readonly<React.PropsWithChildren<ModalProps>>) {
  const [open, setOpen] = React.useState(false)
  const isDesktop = useMediaQuery("(min-width: 768px)")

  if (isDesktop) {
    return (
      <Dialog open={open} onOpenChange={setOpen}>
        <DialogTrigger asChild>{trigger}</DialogTrigger>
        <DialogContent className="sm:max-w-[425px]">
          <DialogHeader>
            <DialogTitle>{title}</DialogTitle>
            <DialogDescription>{description}</DialogDescription>
          </DialogHeader>
          {children}
        </DialogContent>
      </Dialog>
    )
  }

  return (
    <Drawer open={open} onOpenChange={setOpen}>
      <DrawerTrigger asChild>{trigger}</DrawerTrigger>
      <DrawerContent>
        <DrawerHeader>
          <DrawerTitle>{title}</DrawerTitle>
          <DrawerDescription>{description}</DrawerDescription>
        </DrawerHeader>
        <DrawerFooter>{children}</DrawerFooter>
      </DrawerContent>
    </Drawer>
  )
}
