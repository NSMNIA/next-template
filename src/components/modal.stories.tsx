import type { Meta, StoryObj } from "@storybook/react"

import { Modal } from "./modal"

const meta: Meta<typeof Modal> = {
  component: Modal,
  title: "Components/Modal",
}

export default meta

type Story = StoryObj<typeof Modal>

export const Default: Story = {
  args: {
    trigger: <button>Open modal</button>,
    title: "Dit is een test",
    children: (
      <>
        <p>
          Lorem ipsum dolor sit, amet consectetur adipisicing elit. Reiciendis
          iste architecto placeat maxime corporis optio, cum repudiandae,
          quisquam officia sapiente exercitationem quae nam amet, similique
          dolor at doloremque deleniti debitis obcaecati rem tenetur fuga libero
          accusamus. Distinctio porro dolor officia?
        </p>
      </>
    ),
  },
}
