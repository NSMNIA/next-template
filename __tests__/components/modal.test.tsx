/**
 * @jest-environment jsdom
 */
import { render, screen } from "@testing-library/react"

import { Modal } from "@/components/modal"

beforeAll(() => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  })
})

const staticText = {
  title: "This is the title",
  description: "This is the description",
  trigger: "Open Modal",
  content:
    "Here should be a form or something else, but for now it's just a text",
}

describe("Modal", () => {
  it("renders modal and opens", async () => {
    const { baseElement } = render(
      <Modal
        title={staticText.title}
        description={staticText.description}
        trigger={<button>{staticText.trigger}</button>}
      >
        <p>{staticText.content}</p>
      </Modal>
    )

    const button = screen.getByText(staticText.trigger)
    await button.click()

    const title = screen.getByText(staticText.title)
    const description = screen.getByText(staticText.description)

    expect(title).toBeInTheDocument()
    expect(description).toBeInTheDocument()

    expect(baseElement).toMatchSnapshot()
  })
})

it("renders snapshot modal unchanged", () => {
  const { container } = render(
    <Modal
      title={staticText.title}
      description={staticText.description}
      trigger={<button>{staticText.trigger}</button>}
    >
      <p>{staticText.content}</p>
    </Modal>
  )
  expect(container).toMatchSnapshot()
})
