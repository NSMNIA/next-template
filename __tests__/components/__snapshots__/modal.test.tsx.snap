// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders modal and opens 1`] = `
<body
  data-scroll-locked="1"
  style="position: fixed !important; top: 0px; left: 0px; right: 0px; height: auto; pointer-events: none;"
>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  >
    <button
      aria-controls="radix-:r0:"
      aria-expanded="true"
      aria-haspopup="dialog"
      data-state="open"
      type="button"
    >
      Open Modal
    </button>
  </div>
  <div
    aria-hidden="true"
    class="fixed inset-0 z-50 bg-black/80"
    data-aria-hidden="true"
    data-state="open"
    style="pointer-events: auto;"
    vaul-drawer-visible="false"
    vaul-overlay=""
    vaul-snap-points="false"
    vaul-snap-points-overlay="true"
  />
  <div
    aria-describedby="radix-:r2:"
    aria-labelledby="radix-:r1:"
    class="bg-background fixed inset-x-0 bottom-0 z-50 mt-24 flex h-auto flex-col rounded-t-[10px] border"
    data-state="open"
    id="radix-:r0:"
    role="dialog"
    tabindex="-1"
    vaul-drawer=""
    vaul-drawer-direction="bottom"
    vaul-drawer-visible="false"
  >
    <div
      class="bg-muted mx-auto mt-4 h-2 w-[100px] rounded-full"
    />
    <div
      class="grid gap-1.5 p-4 text-center sm:text-left"
    >
      <h2
        class="text-lg font-semibold leading-none tracking-tight"
        id="radix-:r1:"
      >
        This is the title
      </h2>
      <p
        class="text-muted-foreground text-sm"
        id="radix-:r2:"
      >
        This is the description
      </p>
    </div>
    <div
      class="mt-auto flex flex-col gap-2 p-4"
    >
      <p>
        Here should be a form or something else, but for now it's just a text
      </p>
    </div>
  </div>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
</body>
`;

exports[`renders snapshot modal unchanged 1`] = `
<div>
  <button
    aria-controls="radix-:r3:"
    aria-expanded="false"
    aria-haspopup="dialog"
    data-state="closed"
    type="button"
  >
    Open Modal
  </button>
</div>
`;
